File: tools/py/file_finder.py
"""
文件查找工具 - 根据文件名模式查找文件（类似 find）

功能说明：
- 递归遍历目录查找匹配的文件
- 支持通配符模式（如 *.py, config.*.js）
- 可限制返回结果数量
- 纯 Python 实现，无外部依赖

使用方法：
    python tools/py/file_finder.py --pattern "模式" [选项]

参数说明：
    --pattern PATTERN        文件名匹配模式（必need），支持 * 和 ? 通配符
    --path PATH              搜索根目录（默认：当前目录）
    --limit NUM              最大返回结果数（默认：无限制）

输出格式：
    {
      "data": {
        "files": ["文件路径1", "文件路径2", ...]
      },
      "metadata": {
        "elapsed_seconds": 耗时(秒),
        "timeout_threshold": 10,
        "version": "1.1.0"
      }
    }

使用示例：
    # 查找所有 Markdown 文件
    python tools/py/file_finder.py --pattern "*.md" --path ./docs

    # 查找配置文件（限制 10 个结果）
    python tools/py/file_finder.py --pattern "config.*" --limit 10

    # 查找 Python 测试文件
    python tools/py/file_finder.py --pattern "test_*.py"

版本信息：
    版本：1.1.0
    更新日期：2025-12-02
"""

import os
import json
import argparse
import fnmatch
import time

def find_files(path, pattern, limit):
    matches = []
    for root, dirs, files in os.walk(path):
        for filename in fnmatch.filter(files, pattern):
            matches.append(os.path.join(root, filename))
            if limit and len(matches) >= limit:
                return matches
    return matches

def main():
    start_time = time.time()
    
    parser = argparse.ArgumentParser(description="File Finder")
    parser.add_argument("--pattern", required=True, help="Glob pattern")
    parser.add_argument("--path", default=".", help="Root directory to search")
    parser.add_argument("--limit", type=int, help="Max results")
    
    args = parser.parse_args()
    
    files = find_files(args.path, args.pattern, args.limit)
    
    elapsed_time = round(time.time() - start_time, 2)
    result = {
        "data": {"files": files},
        "metadata": {"elapsed_seconds": elapsed_time, "timeout_threshold": 10, "version": "1.1.0"}
    }
    print(json.dumps(result, indent=2, ensure_ascii=False))

if __name__ == "__main__":
    main()

File: tools/js/file_finder.js
/**
 * 文件查找工具 - 根据文件名模式查找文件（类似 find）
 * 
 * 功能说明：
 * - 递归遍历目录查找匹配的文件
 * - 支持通配符模式（如 *.js, config.*.json）
 * - 可限制返回结果数量
 * - 纯 JavaScript 实现，无外部依赖
 * 
 * 使用方法：
 *   node tools/js/file_finder.js --pattern "模式" [选项]
 * 
 * 参数说明：
 *   --pattern PATTERN        文件名匹配模式（必需），支持 * 和 ? 通配符
 *   --path PATH              搜索根目录（默认：当前目录）
 *   --limit NUM              最大返回结果数（默认：无限制）
 * 
 * 输出格式：
 *   {
 *     "data": {
 *       "files": ["文件路径1", "文件路径2", ...]
 *     },
 *     "metadata": {
 *       "elapsed_seconds": 耗时(秒),
 *       "timeout_threshold": 10,
 *       "version": "1.1.0"
 *     }
 *   }
 * 
 * 使用示例：
 *   // 查找所有 Markdown 文件
 *   node tools/js/file_finder.js --pattern "*.md" --path ./docs
 * 
 *   // 查找配置文件（限制 10 个结果）
 *   node tools/js/file_finder.js --pattern "config.*" --limit 10
 * 
 *   // 查找 JavaScript 测试文件
 *   node tools/js/file_finder.js --pattern "test_*.js"
 * 
 * 版本信息：
 *   版本：1.1.0
 *   更新日期：2025-12-02
 */

const fs = require('fs');
const path = require('path');

// Simple glob matcher
function matchGlob(str, pattern) {
    let regexStr = pattern.replace(/[.+^${}()|[\]\\]/g, '\\$&');
    regexStr = regexStr.replace(/\*/g, '.*').replace(/\?/g, '.');
    return new RegExp(`^${regexStr}$`).test(str);
}

function findFiles(rootPath, pattern, limit) {
    const matches = [];
    
    function walk(currentPath) {
        if (limit && matches.length >= limit) return;
        
        let entries;
        try {
            entries = fs.readdirSync(currentPath);
        } catch (e) { return; }
        
        for (const entry of entries) {
            if (limit && matches.length >= limit) return;
            
            const fullPath = path.join(currentPath, entry);
            let stat;
            try {
                stat = fs.statSync(fullPath);
            } catch (e) { continue; }
            
            if (stat.isDirectory()) {
                walk(fullPath);
            } else {
                if (matchGlob(entry, pattern)) {
                    matches.push(fullPath);
                }
            }
        }
    }
    
    walk(rootPath);
    return matches;
}

function main() {
    const startTime = Date.now();
    
    const args = process.argv.slice(2);
    const options = {
        pattern: '',
        path: '.',
        limit: undefined
    };
    
    for (let i = 0; i < args.length; i++) {
        switch (args[i]) {
            case '--pattern': options.pattern = args[++i]; break;
            case '--path': options.path = args[++i]; break;
            case '--limit': options.limit = parseInt(args[++i]); break;
        }
    }
    
    if (!options.pattern) {
        console.error("Error: --pattern is required");
        process.exit(1);
    }
    
    const files = findFiles(options.path, options.pattern, options.limit);
    
    const elapsedTime = ((Date.now() - startTime) / 1000).toFixed(2);
    const result = {
        data: { files: files },
        metadata: {
            elapsed_seconds: parseFloat(elapsedTime),
            timeout_threshold: 10,
            version: "1.1.0"
        }
    };
    console.log(JSON.stringify(result, null, 2));
}

main();
